presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java

```diff
@@ -393,6 +396,15 @@ public TaskInfo abortTaskResults(TaskId taskId, OutputBufferId bufferId)
         return tasks.getUnchecked(taskId).abortTaskResults(bufferId);
     }
 
+    @Override
+    public void removeRemoteSource(TaskId taskId, TaskId remoteSourceTaskId)
+    {
+        requireNonNull(taskId, "taskId is null");
+        requireNonNull(remoteSourceTaskId, "srcTaskId is null");
+
+        tasks.getUnchecked(taskId).removeRemoteSource(remoteSourceTaskId);
```

Comment:
@arhimondr : I see there could be difference to `abortTaskResults`, `cancelTask`... since the call is initialized by children failed stages, in theory this could happen.  Let me just add a check and make it log.

Reply:
@arhimondr : Actually `tasks` is a `LoadingCache`. So if the task is not created, it will create on the fly :).

ExchangeClient might not created yet, though. I think we will just fail the query for such (unlucky) timing for now.

Reply:
> ExchangeClient might not created yet, though. I think we will just fail the query for such (unlucky) timing for now.

Yeah, that was the case i pointed out. Yes, i agree. Let's forget about that for now. There will be a second line of defence anyway - whole stage retries .

Reply:
@arhimondr : We can add warning in `SqlTask#removeRemoteSource` if there is no ExchangeClient :)  . Again I defer it as future work ;)